ampd/src/tofnd/client.rs

```diff
@@ -2,156 +2,185 @@
+}
+
+pub struct Client {
+    party_uid: String,
+    sender: mpsc::Sender<Request>,
 }
 
 #[async_trait]
-impl MultisigClient for multisig_client::MultisigClient<Channel> {
-    async fn keygen(&mut self, request: KeygenRequest) -> Result<KeygenResponse, Status> {
-        self.keygen(request)
-            .await
-            .and_then(|response| {
-                response
-                    .into_inner()
-                    .keygen_response
-                    .ok_or_else(|| Status::internal("keygen response is empty"))
-            })
-            .into_report()
+impl Multisig for Client {
+    async fn keygen(&self, key_uid: &str) -> Result<PublicKey> {
+        self.send(|handle| Request::Keygen {
+            request: KeygenRequest {
+                key_uid: key_uid.to_string(),
+                party_uid: self.party_uid.clone(),
+            },
+            handle,
+        })
+        .await
     }
 
-    async fn sign(&mut self, request: SignRequest) -> Result<SignResponse, Status> {
-        self.sign(request)
-            .await
-            .and_then(|response| {
-                response
-                    .into_inner()
-                    .sign_response
-                    .ok_or_else(|| Status::internal("sign response is empty"))
-            })
-            .into_report()
+    async fn sign(&self, key_uid: &str, data: Vec<u8>, pub_key: &PublicKey) -> Result<Signature> {
+        self.send(|handle| Request::Sign {
+            request: SignRequest {
+                key_uid: key_uid.to_string(),
+                msg_to_sign: data,
+                party_uid: self.party_uid.clone(),
+                pub_key: pub_key.to_bytes(),
+            },
+            handle,
+        })
+        .await
     }
+}
+
+impl Client {
+    async fn send<T>(&self, with_handle: impl FnOnce(Handle<T>) -> Request) -> Result<T> {
+        let (tx, rx) = oneshot::channel();
 
-    async fn key_presence(
-        &mut self,
-        request: KeyPresenceRequest,
-    ) -> Result<KeyPresenceEnum, Status> {
-        self.key_presence(request)
+        self.sender
+            .send(with_handle(tx))
             .await
-            .map(Response::into_inner)
-            .and_then(|response| {
-                KeyPresenceEnum::from_i32(response.response)
-                    .ok_or_else(|| Status::internal("invalid key presence response"))
-            })
             .into_report()
+            .change_context(Error::SendFailed)?;
+
+        rx.await.into_report().change_context(Error::RecvFailed)?
     }
 }
 
-pub struct Client<T: MultisigClient> {
-    client: T,
+pub struct TofndClient<T: MultisigClient> {
     party_uid: String,
+    client: T,
+    channel: (mpsc::Sender<Request>, mpsc::Receiver<Request>),
 }
 
-impl<T: MultisigClient> Client<T> {
-    pub fn new(client: T, party_uid: String) -> Self {
-        Self { client, party_uid }
+impl<T: MultisigClient> TofndClient<T> {
```

Comment:
I kinda wanna get rid of the word "client" here. There are way too many things call clients in the tofnd mod, and they don't always mean the same thing. @cgorenflo any idea?